content/posts/2022-08-14-libc++-format.md

```diff
@@ -0,0 +1,273 @@
+    A | U+0041 { LATIN CAPITAL LETTER A }
+    Á | U+00C1 { LATIN CAPITAL LETTER A WITH ACUTE }
+    Á | U+0041 U+0301 { LATIN CAPITAL LETTER A } { COMBINING ACUTE ACCENT }
+    Ĳ | U+0132 { LATIN CAPITAL LIGATURE IJ }
+    Δ | U+0394 { GREEK CAPITAL LETTER DELTA }
+    Щ | U+0429 { CYRILLIC CAPITAL LETTER SHCHA }
+    א | U+05D0 { HEBREW LETTER ALEF }
+    ش | U+0634 { ARABIC LETTER SHEEN }
+   〉 | U+3009 { RIGHT-POINTING ANGLE BRACKET }
+   界 | U+754C { CJK Unified Ideograph-754C }
+   🦄 | U+1F921 { UNICORN FACE }
+   👨‍👩‍👧‍👦 | U+1F468 U+200D U+1F469 U+200D U+1F467 U+200D U+1F466 { Family: Man, Woman, Girl, Boy } 
+```
+
+A huge benefit of ``std::format`` over ``printf`` is its type safety.
+When attempting to format a string as a number, it will result in a compilation
+failure.
+```cpp
+int main() {
+  std::cout << std::format("{0:#08B}, {0:#08o}, {0:08}, {0:#08X}", "15");
+}
+
+```
+The compiler output starts with this error, followed by a lot of not too useful messages.
+```
+error: call to consteval function 'std::basic_format_string<char, const char (&)[3]>::basic_format_string<char[37]>' is not a constant expression
+  std::cout << std::format("{0:#08B}, {0:#08o}, {0:08}, {0:#08X}", "15");
+```
+
+Next to formatting to a string it is possible to format:
+
+  -  to an output iterator,
+```cpp
+int main() {
+  std::format_to(
+    std::ostream_iterator<char>(std::cout, ""),
+    "Hello {} in C++{}\n", "std::format", 20);
+}
+```
+  - determine the output size,
+
+```cpp
+int main() {
+  std::cout << std::formatted_size("Hello {} in C++{}\n", "std::format", 20);
+}
+```
+```
+27
+```
+ - or limit the size of the output.
+```cpp
+int main() {
+  std::format_to(
+    std::ostream_iterator<char>(std::cout, ""),
+    11, "Hello {} in C++{}\n", "std::format", 20);
+}
+```
+```
+Hello std
+```
+
+An example of formatting user defined types is available in the standard
+library. It has formatting support for the ``chrono`` library. (This is not
+available in libc++ yet.)
+
+More examples and details of the specification can be found in this 
+[{fmt} cheet sheet](https://hackingcpp.com/cpp/libs/fmt.html).
+
+
+## The status in libc++15
+
+In libc++15 most of the basic text formatting is complete. All major papers
+have been implemented, but some defect reports are not implemented. Another thing
+the libc++ team wants to work on are performance improvements. Some of these
+improvements have already landed for libc++16, others are planned.  The
+improvements may affect the ABI. Next to our planned changes the C++ committee
+may also vote in an ABI breaking change. The compile-time errors when using an
+invalid format string are not great. This is something which we want to improve
+in the future.
+
+Since the library is not entirely complete, nor ABI stable, it is shipped in
+the experimental library. To use the code in libc++ you need to compile the
+code like:
```

Comment:
```suggestion
Since the library is not entirely complete and the ABI may need to change (due to planned improvements but also changes voted by the C++ Committee), it is shipped as an experimental feature in LLVM 15. To use the code in libc++ you need to compile the code like:
```